File: backend/app/modules/catalogue/schemas/product.py
from datetime import datetime
from typing import Annotated
from uuid import UUID

from pydantic import BaseModel, ConfigDict, StringConstraints, model_validator

TrimmedName = Annotated[
    str,
    StringConstraints(
        strip_whitespace=True,
        min_length=1,
        max_length=200,
    ),
]


class ProductCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: TrimmedName
    description: str | None = None
    specifications: dict[str, object] | None = None
    category_id: UUID


class ProductResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    product_id: UUID
    name: str
    description: str | None
    specifications: dict[str, object] | None
    is_active: bool
    created_at: datetime
    updated_at: datetime
    category_id: UUID
    slug: str


class ProductUpdate(BaseModel):
    model_config = ConfigDict(from_attributes=True, extra="forbid")

    name: TrimmedName | None = None
    description: str | None = None
    specifications: dict[str, object] | None = None
    is_active: bool | None = None
    category_id: UUID | None = None

    @model_validator(mode="after")
    def require_atleast_one_field(self):
        if not self.model_fields_set:
            raise ValueError("Atleast one field must be provided")

        non_nullable_fields = ["name", "is_active", "category_id"]
        for field in non_nullable_fields:
            if field in self.model_fields_set and getattr(self, field) is None:
                raise ValueError(f"{field} cannot be null")

        return self


class ProductListResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    items: list[ProductResponse]
    total: int
    page: int
    page_size: int

File: apps/admin-web/src/features/products/types/product.ts




export type Product = {
    product_id: string;
    category_id: string;
    name: string;
    description: string | null;
    slug: string;
    specifications: Record<string, unknown> | null;
    is_active: boolean;
    created_at: string;
    updated_at: string;
}

export type ProductListResponse = {
    items: Product[];
    total: number;
    page: number;
    page_size: number;
}

export type ProductListParams = {
    page: number;
    page_size: number;
    search?: string;
    category_id?: string;
    is_active?: boolean;
}


export type ProductCreateInput = {
    name: string;
    description?: string | null;
    category_id: string;
}

export type ProductUpdateInput = {
    name?: string;
    description?: string | null;
    category_id?: string;
    is_active?: boolean;
    
}

